gateway/api_definition.go

```diff
@@ -988,14 +988,14 @@ func (a *APISpec) getURLStatus(stat URLStatus) RequestStatus {
 // URLAllowedAndIgnored checks if a url is allowed and ignored.
 func (a *APISpec) URLAllowedAndIgnored(r *http.Request, rxPaths []URLSpec, whiteListStatus bool) (RequestStatus, interface{}) {
 	// Check if ignored
-	for _, v := range rxPaths {
-		if !v.Spec.MatchString(r.URL.Path) {
+	for i := range rxPaths {
+		if !rxPaths[i].Spec.MatchString(r.URL.Path) {
```

Comment:
this is quite interesting optimization - so access by index is faster than just sequential iteration over array?

Reply:
would be interesting to see `BenchmarkVersioning` cmp-results vs master

Reply:
It's not just sequential access over array - before you are making the allocation to `v`.

Master: `list UrlAllowedAndIgnored`
```
(pprof) list URLAllowedAndIgnored
Total: 3.43MB
ROUTINE ======================== github.com/TykTechnologies/tyk/gateway.(*APISpec).URLAllowedAndIgnored in /Users/ahmet/go/src/github.com/TykTechnologies/tyk/gateway/api_definition.go
       7kB        7kB (flat, cum)   0.2% of Total
         .          .    986:}
         .          .    987:
         .          .    988:// URLAllowedAndIgnored checks if a url is allowed and ignored.
         .          .    989:func (a *APISpec) URLAllowedAndIgnored(r *http.Request, rxPaths []URLSpec, whiteListStatus bool) (RequestStatus, interface{}) {
         .          .    990:	// Check if ignored
       7kB        7kB    991:	for _, v := range rxPaths {
         .          .    992:		if !v.Spec.MatchString(r.URL.Path) {```
```

New: it doesn't even show, because we are not allocating to `v`.

Reply:
But rxPath[i] is like v
That feels like it’s the optimiser job

Reply:
This was done using a large `[]int`. But rxPaths is considerably larger.

```
BenchmarkRangeCopy-8   	     490	   2497576 ns/op	     282 B/op	       3 allocs/op
BenchmarkRangeIndex-8   	     470	   2476958 ns/op	     262 B/op	       2 allocs/op
```

Reply:
@dencoded 

```
OLD: BenchmarkDefaultVersion-8   	    1233	   1029221 ns/op	  226285 B/op	    1527 allocs/op
NEW: BenchmarkDefaultVersion-8   	    1336	    815036 ns/op	  155331 B/op	    1507 allocs/op
```

Not created specific benchmark - but here is an example bench which touches `URLAllowedAndIgnored`